plugins/flashrom/fu-plugin-flashrom.c

```diff
@@ -0,0 +1,198 @@
+/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
+ *
+ * Copyright (C) 2017 Richard Hughes <richard@hughsie.com>
+ *
+ * Licensed under the GNU General Public License Version 2
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "config.h"
+
+#include "fu-plugin.h"
+#include "fu-plugin-vfuncs.h"
+
+struct FuPluginData {
+	gchar			*flashrom_fn;
+};
+
+void
+fu_plugin_init (FuPlugin *plugin)
+{
+	fu_plugin_alloc_data (plugin, sizeof (FuPluginData));
+}
+
+void
+fu_plugin_destroy (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_free (data->flashrom_fn);
+}
+
+
+static void
+fu_plugin_flashrom_read_cb (const gchar *line, gpointer user_data)
+{
+	FuPlugin *plugin = FU_PLUGIN (user_data);
+	if (g_strcmp0 (line, "Reading flash...") == 0)
+		fu_plugin_set_status (plugin, FWUPD_STATUS_DEVICE_VERIFY);
+	//FIXME get percentage completion...
+	g_debug ("got '%s'", line);
+}
+
+static void
+fu_plugin_flashrom_write_cb (const gchar *line, gpointer user_data)
+{
+	FuPlugin *plugin = FU_PLUGIN (user_data);
+	if (g_strcmp0 (line, "Writing flash...") == 0)
+		fu_plugin_set_status (plugin, FWUPD_STATUS_DEVICE_WRITE);
+	//FIXME get percentage completion...
+	g_debug ("got '%s'", line);
+}
+
+static FuDevice *
+fu_plugin_flashrom_create_device (FuPlugin *plugin,
+				  const gchar *guid,
+				  const gchar *id,
+				  GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_autofree gchar *device_id = g_strdup_printf ("flashrom-%s", id);
+	g_autofree gchar *firmware_orig = NULL;
+	g_autofree gchar *basename = NULL;
+	g_autoptr(FuDevice) dev = fu_device_new ();
+
+	/* if the original firmware doesn't exist, grab it now */
+	basename = g_strdup_printf ("flashrom-%s.bin", id);
+	firmware_orig = g_build_filename (LOCALSTATEDIR, "lib", "fwupd",
+					  "builder", basename, NULL);
```

Comment:
It's a little unclear to me why this needs to be dumped at startup.  It's not used at runtime, and nothing else is done with it earlier on.  So that would be slowing down fwupd startup for not a lot of benefit?

I would think it makes more sense to dump during update instead.

Reply:
After later realizing that the `bubblewrap `magic happens *before* update, then maybe this should live in `update_prepare` instead of during startup.

Reply:
100% correct, this should of course be in `update_prepare` -- fixed.